File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int[] left = new int[heights.length];
        int[] right = new int[heights.length];

        findNextSmallestRight(heights, right);
        findNextSmallestLeft(heights, left);

        int max = 0;

        for(int i = 0; i < heights.length; i++) {
            int area = ((right[i]-1)-(left[i]+1)+1)*heights[i];
            max = Math.max(max,area);
        }

        return max;

    }

    private void findNextSmallestRight(int[] heights, int[] right) {
        Stack<Integer> stack = new Stack<>();
        for(int i = 0; i < heights.length; i++) {
            while(!stack.isEmpty() && heights[stack.peek()] > heights[i]) {
                right[stack.pop()] = i;
            }
            stack.push(i);
        }

        while(!stack.isEmpty()) {
            right[stack.pop()] = heights.length;
        }
    }

    private void findNextSmallestLeft(int[] heights, int[] left) {
        Stack<Integer> stack = new Stack<>();
        for(int i = heights.length-1; i>=0; i--) {
            while(!stack.isEmpty() && heights[stack.peek()] > heights[i]) {
                left[stack.pop()] = i;
            }
            stack.push(i);
        }

        while(!stack.isEmpty()) {
            left[stack.pop()] = -1;
        }
    }
}

File: Data Structures & Algorithms/longest-palindromic-substring/submission-2.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        max_length = 0
        longest_palindrome = ""

        for i in range(0,len(s)):
            left = i
            right = i
            while left >=0 and right < len(s) and s[left] == s[right]:
                length = right-left+1
                if(length > max_length):
                    max_length = length
                    longest_palindrome = s[left:right+1]
                left -= 1
                right += 1  

        for i in range(1,len(s)):
            left = i
            right = i-1
            while left >=0 and right < len(s) and s[left] == s[right]:
                length = right-left+1
                if(length > max_length):
                    max_length = length
                    longest_palindrome = s[left:right+1]
                left -= 1
                right += 1   

        return  longest_palindrome            
                
              
        